proof/invariant-abstract/AARCH64/ArchInvariants_AI.thy

```diff
@@ -2198,73 +2188,42 @@ lemma is_aligned_ptrFromPAddr:
+  "\<lbrakk> pspace_aligned s; pts_of s pt_ptr = Some pt \<rbrakk> \<Longrightarrow> is_aligned pt_ptr (pt_bits (pt_type pt))"
   by (fastforce dest: pspace_alignedD simp: in_omonad bit_simps)
 
 lemma is_aligned_pt_slot_offset_pte:
-  "is_aligned pt (pt_bits vsp) \<Longrightarrow> is_aligned (pt_slot_offset level pt vref) pte_bits"
+  "is_aligned pt (pt_bits pt_t) \<Longrightarrow> is_aligned (pt_slot_offset level pt vref) pte_bits"
   unfolding pt_slot_offset_def
   by (simp add: is_aligned_add bit_simps is_aligned_weaken is_aligned_shift)
 
-(* Show that VSRootPTs are distinct from any index ptes_of might compute for a NormalPT *)
-lemma vs_root_pt_idx_distinct:
-  "\<lbrakk> pts_of s pt = Some (VSRootPT vs);
-     pts_of s (table_base False (pt_slot_offset max_pt_level pt vref)) = Some (NormalPT npt);
-     pspace_aligned s; pspace_distinct s \<rbrakk>
-   \<Longrightarrow> False"
-  apply (clarsimp simp: in_omonad)
-  apply (prop_tac "pt \<noteq> table_base False (pt_slot_offset max_pt_level pt vref)", clarsimp)
-  apply (frule_tac y="table_base False (pt_slot_offset max_pt_level pt vref)" in pspace_distinctD;
-         assumption?)
-  apply (drule (1) pspace_alignedD)+
-  apply (cases "config_ARM_PA_SIZE_BITS_40")
-   apply (clarsimp simp: pt_slot_offset_def bit_simps pt_index_def pt_bits_left_def size_max_pt_level
-                   split: if_split_asm)
-   apply (simp add: is_aligned_no_overflow_mask)
-   apply (subst (asm) word_plus_and_or_coroll, word_eqI)+
-   apply word_bitwise
-   apply simp
-  (* This is the case where all tables are equal size, which should be the easier one, but the
-     carefully crafted simp rules don't apply at this level, so first establish the right context *)
-  apply (prop_tac "table_base False (pt_slot_offset max_pt_level pt vref) =
-                   table_base True (pt_slot_offset max_pt_level pt vref)", simp add: bit_simps)
-  apply (prop_tac "table_size False = table_size True", simp add: bit_simps)
-  apply clarsimp
-  apply (subst (asm) table_base_pt_slot_offset; simp?)
-  apply (simp add: pt_bits_def)
-  done
-
 lemma pt_slot_offset_pt_range:
-  "\<lbrakk> ptes_of s (pt_slot_offset level pt vref) = Some pte; level \<le> max_pt_level;
-     is_VSRootPT ptable = (level = max_pt_level);
-     pts_of s pt = Some ptable; pspace_aligned s; pspace_distinct s \<rbrakk>
+  "\<lbrakk> ptes_of s level (pt_slot_offset level pt vref) = Some pte;
+     pts_of s pt = Some ptable; is_aligned pt (pt_bits level) \<rbrakk>
    \<Longrightarrow> pte \<in> pt_range ptable"
-  apply (clarsimp simp: ptes_of_Some)
-  apply (frule (1) pspace_aligned_pts_ofD)
-  apply (cases "level = max_pt_level"; clarsimp simp: is_VSRootPT_def)
-  apply (cases "\<exists>npt. pts_of s (table_base False (pt_slot_offset max_pt_level pt vref)) =
-                      Some (NormalPT npt)"; clarsimp)
-  apply (drule (3) vs_root_pt_idx_distinct)
-  apply simp
-  done
+  for level::vm_level
```

Comment:
I don't think I've seen this, either. I'd try to investigate this myself, but searching for "for" in the documentation was of course not the most useful thing to do...

Reply:
This is yet another way to state structured goals (in addition to `fixes/assumes/shows` and `/\ .. ==> ..`. It's been introduced in Isabelle2016, but we don't tend to use it much apart from the `for` form which is nice for constraining types without forcing the `assumes/shows` form which needs more space.

The general form is `P if A and B for x :: T`, which translates to `/\x::T. [| A; B |] ==> P`.

Reply:
I thought it must have been something like that, but seeing that it was out of the quotation marks for the statement, I thought it might have been doing something to the proof. Thanks for the explanations, though!